runtime/compute/cker/include/cker/Shape.h

```diff
@@ -56,136 +69,171 @@ class Shape
     }
   }
 
+  // Constructor that creates a shape from an array of dimension data.
   Shape(int dimensions_count, const int32_t *dims_data) : _size(0)
   {
     ReplaceWith(dimensions_count, dims_data);
   }
 
+  // Initializer list constructor.
+  // Marked explicit to avoid unintended overload resolution.
   Shape(const std::initializer_list<int> init_list) : _size(0) { BuildFrom(init_list); }
 
-  // Avoid using this constructor.  We should be able to delete it when C++17
-  // rolls out.
-  Shape(Shape const &other) : _size(other.DimensionsCount())
+  // Copy constructor
+  Shape(const Shape &other) : _size(other._size)
   {
-    if (_size > kMaxSmallSize)
+    if (_size <= kMaxSmallSize)
+    {
+      // When the number of dimensions is small, copy the fixed array.
+      dims_ = std::get<std::array<int32_t, kMaxSmallSize>>(other.dims_);
+    }
+    else
     {
-      _dims_pointer = new int32_t[_size];
+      // Otherwise, copy the dynamically allocated vector.
+      dims_ = std::get<std::vector<int32_t>>(other.dims_);
     }
-    std::memcpy(DimsData(), other.DimsData(), sizeof(int32_t) * _size);
   }
+  Shape(Shape &&other) = default;
```

Comment:
I enabled Move Semantics deleted by the custom copy constructor:
It allows instances of `Shape` to be moved efficiently. Instead of copying potentially large amounts of data (especially for dynamically allocated storage when dimensions exceed `kMaxSmallSize`), the object’s resources can be transferred (or "moved") to a new object with minimal overhead.

For Performance Benefits:
In modern C++ (C++11 and later), move semantics are essential to optimize performance and resource management. By marking the move constructor as default, we ensure that move operations are available and implemented in the most efficient way possible by the compiler.